Introduce `ProtocolParameters` type.

jonathanknowles opened the pull request:
# Related Issue

#1693 

# Overview

This PR prepares the ground for reporting the live value of the **_d_** protocol parameter.

We introduce the following type to hold **protocol parameters**:

```hs
data ProtocolParameters = ProtocolParameters  
    { decentralizationLevel
        :: Quantity "percent" Percentage      
    , txParameters
        :: TxParameters       
    }
```

Using the above type, we can rewrite the following type:

```diff
- data GenesisBlockParameters = GenesisBlockParameters
-     { staticParameters :: BlockchainParameters
-     ,     txParameters ::         TxParameters
-     }
+ data NetworkParameters = NetworkParameters
+     {  genesisParameters ::  GenesisParameters
+     , protocolParameters :: ProtocolParameters
+     }
```

Finally, we change the type of `toApiNetworkParameters` to:
```diff
- toApiNetworkParameters :: BlockchainParameters -> ApiNetworkParameters
+ toApiNetworkParameters ::    NetworkParameters -> ApiNetworkParameters
```

## Outstanding Work

- [x] Ensure that related functions and values are named consistently.

jonathanknowles committed:
Rename `GenesisBlockParameters` to `NetworkParameters`.

jonathanknowles committed:
Rename `BlockchainParameters` to `GenesisParameters`.

jonathanknowles committed:
Rename `genesisParameters` to `dummyGenesisParameters`.

jonathanknowles committed:
Rename `genesisBlockParameters` to `dummyNetworkParameters`.

jonathanknowles committed:
Rename `genesisTxParameters` to `dummyTxParameters`.

jonathanknowles committed:
Rename `staticParameters` to `genesisParameters`.

jonathanknowles committed:
Adjust formatting to meet coding standards.

jonathanknowles committed:
Construct `ApiNetworkParameters` from `NetworkParameters`.

jonathanknowles committed:
Introduce type `ProtocolParameters`.

rvl commented on lib/byron/src/Cardano/Wallet/Byron/Network.hs:
```diff
@@ -231,17 +233,19 @@ withNetworkLayer tr gbp addrInfo versionData action = do
-        let points = reverse $ genesisPoint : (toPoint getGenesisBlockHash getEpochLength <$> headers)
+        link =<< async
+            (connectClient tr handlers (const client) versionData addrInfo)
+        let points = reverse $ genesisPoint :
```
Since you're reformatting - I prefer `reverse (x:xs)` to `reverse $ x : xs`.

jonathanknowles replied:
My reason for using `$` here is to create a suitable point for a line break.

I'd be happy to use brackets instead. But is there a way to do this without violating the coding standards?

rvl commented on lib/byron/test/unit/Cardano/Wallet/Byron/NetworkSpec.hs:
```diff
@@ -41,20 +45,23 @@ spec = describe "getTxParameters" $ do
-            let gbp' = gbp { txParameters = zeroTxParameters }
+            let pps = protocolParameters gbp
+            let pps' = pps { txParameters = zeroTxParameters }
+            let gbp' = gbp { protocolParameters = pps' }
```
A lens operator might look better here.

jonathanknowles replied:
Fixed in 0fc13db89eec38e41f21166d3555197a435699c2.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -83,9 +83,10 @@ module Cardano.Wallet.Primitive.Types
-    -- * BlockchainParameters
-    , GenesisBlockParameters (..)
-    , BlockchainParameters (..)
+    -- * GenesisParameters
```
```suggestion
    -- * Blockchain Parameters
```

jonathanknowles replied:
I think that "Network Parameters" probably makes more sense here, as `NetworkParameters` is the type that sits at the root of the tree, so to speak. (`NetworkParameters` includes both `GenesisParameters` and `ProtocolParameters`.)

jonathanknowles replied:
Fixed in 39b079cb6319a4dcb8c2ffea0c1bd9232f100717.

rvl reviewed (approved):
Looks fine. These names are better I think. I was getting bored towards the end of the diff though.
You might want to rename the variables too, otherwise it could get confusing. E.g. "gbp" was a mnemonic for GenesisBlockParameters. But you may wish to merge sooner rather than later to avoid possible merge conflicts.

jonathanknowles committed:
Use lenses to simplify record update.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1700#discussion_r432351837

jonathanknowles committed:
Update Haddock comment to reflect `NetworkParameters` grouping.

jonathanknowles committed:
Rename `gbp` to `np` where appropriate.

The identifier `gbp` was often used for values of type
`GenesisBlockParameters`.

Since `GenesisBlockParameters` has been renamed to `NetworkParameters`,
it makes sense to also rename `gbp` to `np`.

jonathanknowles committed:
Rename `bp` to `gp` where appropriate.

The identifier `bp` was often used to represent values of type
`BlockchainParameters`.

However, this type has been renamed to `GenesisParameters`. It therefore
also makes sense to rename `bp` to `gp`, where appropriate.

jonathanknowles committed:
Rename `mainnetBlockchainParameters` to `mainnetNetworkParameters`.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -499,28 +503,18 @@ data ApiNetworkParameters = ApiNetworkParameters
-    (Quantity $ unEpochLength $ getEpochLength bp)
-    (getEpochStability bp)
+toApiNetworkParameters :: NetworkParameters -> ApiNetworkParameters
+toApiNetworkParameters (NetworkParameters gp pps) = ApiNetworkParameters
```
`pps` should probably be renamed to `pp`.

We generally don't append the `s` suffix to identifiers that represent values of types ending in `Parameters`.

jonathanknowles replied:
Fixed in 3ac0a59db5085f5aea98cdc26b0441b7fa82cf31.

jonathanknowles committed:
Rename `pps` to `pp`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1700/files#r433709423

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -1283,22 +1284,22 @@ computeUtxoStatistics btype utxos =
-instance Buildable GenesisBlockParameters where
-    build (GenesisBlockParameters bp txp) = build bp <> build txp
+instance Buildable NetworkParameters where
+    build (NetworkParameters gp txp) = build gp <> build txp
```
```suggestion
    build (NetworkParameters gp pp) = build gp <> build pp
```

jonathanknowles replied:
Fixed in 70e781f34f8f0bd9f8b954052be4d680b03755a8.

jonathanknowles committed:
Use `pp` to represent value of type `ProtocolParameters`.

jonathanknowles commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs:
```diff
@@ -267,16 +268,16 @@ newNetworkLayer
     -> BaseUrl
     -> Hash "Genesis"
-    -> ExceptT ErrGetBlockchainParams IO ((J.Block, GenesisBlockParameters), NetworkLayer IO t J.Block)
+    -> ExceptT ErrGetBlockchainParams IO ((J.Block, NetworkParameters), NetworkLayer IO t J.Block)
```
```suggestion
    -> ExceptT ErrGetBlockchainParams IO
        ((J.Block, NetworkParameters), NetworkLayer IO t J.Block)
```

jonathanknowles replied:
Fixed in 017aceaa92f96eb6db6e873427a2ff7e014a3b27.

jonathanknowles committed:
Adjust formatting to meet coding standards.

Fixes:

https://github.com/input-output-hk/cardano-wallet/pull/1700/files#r433718304

jonathanknowles committed:
Add reminder to fetch live value of decentralization level.
